### PR TITLE
Support DataFrame.collect for Ballista DataFrames

### LINKED ISSUE #787: Support DataFrame.collect for Ballista DataFrames
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
A DataFrame from a BallistaContext should work just like a DataFrame from a DataFusion ExecutionContext. The user should be able to directly call execution methods such as collect.

**Describe the solution you'd like**
A DataFrame from a BallistaContext should work just like a DataFrame from a DataFusion ExecutionContext. The user should be able to directly call execution methods such as collect.

**Describe alternatives you've considered**
None

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes #787.

 # Rationale for this change

A DataFrame from a BallistaContext should work just like a DataFrame from a DataFusion ExecutionContext. The user should be able to directly call execution methods such as `collect`.

# What changes are included in this PR?

DataFrames created by Ballista have a custom query planner that can create Ballista physical plans.

# Are there any user-facing changes?

Yes - the user no longer needs to call `BallistaContext::collect`.

### CODE REVIEW

**Comment by alamb** [comment]
I will plan to review this tomorrow

**Review by houqp — APPROVED** [review]

File: ballista-examples/src/bin/ballista-dataframe.rs [hunk 1]
```diff
@@ -38,18 +38,7 @@ async fn main() -> Result<()> {
...
-    // logical plan
-    let mut stream = ctx.collect(&df.to_logical_plan()).await?;
-
-    // print the results
-    let mut results = vec![];
-    while let Some(batch) = stream.next().await {
-        let batch = batch?;
-        results.push(batch);
-    }
+    let results = df.collect().await?;
```
Comment (alamb) [inline ballista-examples/src/bin/ballista-dataframe.rs:41]: That is certainly nicer 👍

File: ballista/rust/core/src/execution_plans/distributed_query.rs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        Self {
+            scheduler_url,
+            config,
+            plan,
+        }
+    }
+}
+
+#[async_trait]
+impl ExecutionPlan for DistributedQueryExec {
```
Comment (alamb) [inline ballista/rust/core/src/execution_plans/distributed_query.rs:70]: This is a cool pattern 👍

**Review by alamb — APPROVED** [review]
This is a really nice structure 👍

**Comment by andygrove** [comment]
> This is a really nice structure 👍

Thanks. It was nice that I didn't have to make any changes to DataFusion to make this work.

**Pull request merged**